File: apps/nmp-gallery/ios/NmpGallery/Gallery/ContentComponentPages.swift
import SwiftUI

/// Shared chrome for content-component pages.
private struct ContentPageFrame<Content: View>: View {
    let caption: String
    @ViewBuilder var content: () -> Content

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(caption)
                .font(.caption)
                .foregroundStyle(.secondary)
            VStack(alignment: .leading) {
                content()
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(20)
            .background(Color(.secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 12))
        }
    }
}

/// Reusable toggle shown at the top of pages that render inline mentions.
/// When on, `NostrContentView` receives the raw wire URI; when off, the
/// kernel-backed display name (or truncated pubkey while loading).
private struct RawToggle: View {
    @Binding var rawMode: Bool

    var body: some View {
        HStack {
            Text(rawMode ? "Raw wire" : "Profile")
                .font(.subheadline)
                .foregroundStyle(.secondary)
            Spacer()
            Toggle("", isOn: $rawMode)
                .labelsHidden()
        }
    }
}

// MARK: - Showcase references

/// A reusable rich `ContentTreeWire` exercise. Mirrors the registry's
/// `NostrContentViewPreview` arena and uses the same relay-backed references
/// as the TUI and Android galleries.
private enum ShowcaseContent {
    static var richTree: ContentTreeWire {
        // Arena layout:
        //   0  text "relay note "
        //   1  mention(SHOWCASE_PUBKEY_HEX)
        //   2  text " "
        //   3  eventRef(SHOWCASE_NOTE_NEVENT)
        //   4  paragraph(children: [0,1,2,3])
        return ContentTreeWire(
            nodes: [
                .text("relay note "),
                .mention(
                    NostrWireUri(
                        uri: "nostr:\(SHOWCASE_NPUB)",
                        kind: .profile,
                        primaryId: SHOWCASE_PUBKEY_HEX
                    )
                ),
                .text(" "),
                .eventRef(showcaseNoteURI),
                .paragraph(children: [0, 1, 2, 3]),
            ],
            roots: [4],
            mode: nil
        )
    }
}

// MARK: - content-core

struct ContentCorePage: View {
    var body: some View {
        VStack(spacing: 16) {
            ContentPageFrame(caption: "ContentTreeWire — arena dump") {
                let tree = ShowcaseContent.richTree
                Text("nodes: \(tree.nodes.count)   roots: \(tree.roots.count)")
                    .font(.caption.monospaced())
                    .foregroundStyle(.secondary)
                Text("Render it with NostrContentView; the wire is just data.")
                    .font(.callout)
            }
            ContentPageFrame(caption: "NostrIdenticon.identiconView(forPubkey:size:)") {
                HStack(spacing: 16) {
                    NostrIdenticon.identiconView(forPubkey: SHOWCASE_PUBKEY_HEX, size: 32)
                    NostrIdenticon.identiconView(forPubkey: SHOWCASE_PUBKEY_HEX, size: 40)
                    NostrIdenticon.identiconView(forPubkey: SHOWCASE_PUBKEY_HEX, size: 48)
                    NostrIdenticon.identiconView(forPubkey: SHOWCASE_PUBKEY_HEX, size: 56)
                }
            }
        }
    }
}

// MARK: - content-view

struct ContentViewPage: View {
    @Environment(GalleryModel.self) private var model
    @State private var rawMode = false

    var body: some View {
        VStack(spacing: 16) {
            // A hidden avatar owns the claim for the showcase pubkey while this
            // page is visible — same mechanism as `ProfileEmbedPage`, so
            // `model.profile(forPubkey:)` resolves to the display name instead
            // of the truncated hex fallback.
            NostrAvatar(pubkey: SHOWCASE_PUBKEY_HEX, size: 0)
                .frame(width: 0, height: 0)
                .clipped()
            RawToggle(rawMode: $rawMode)
            ContentPageFrame(caption: "NostrContentView(tree:)") {
                NostrContentView(
                    tree: ShowcaseContent.richTree,
                    mentionLabel: { uri in
                        rawMode
                            ? uri.uri
                            : model.profile(forPubkey: uri.primaryId)?.displayName
                                ?? NostrContentView.defaultMentionLabel(uri)
                    }
                )
            }
        }
    }
}

// MARK: - content-mention-chip

/// Inline-mention showcase: a `ContentTreeWire` that renders "Hey @pablof7z,
/// how are you?". The mention node's `primaryId` is `SHOWCASE_PUBKEY_HEX`
/// and the `mentionLabel` closure looks up the live kind:0 profile the
/// kernel claimed for the mounted component. Raw toggle shows the wire URI vs the
/// resolved display name.
private enum MentionShowcase {
    /// Arena layout:
    ///   0  text "Hey "
    ///   1  mention(SHOWCASE_PUBKEY_HEX)
    ///   2  text ", how are you?"
    ///   3  paragraph(children: [0, 1, 2])
    static var note: ContentTreeWire {
        ContentTreeWire(
            nodes: [
                .text("Hey "),
                .mention(
                    NostrWireUri(
                        uri: "nostr:\(SHOWCASE_NPUB)",
                        kind: .profile,
                        primaryId: SHOWCASE_PUBKEY_HEX
                    )
                ),
                .text(", how are you?"),
                .paragraph(children: [0, 1, 2]),
            ],
            roots: [3],
            mode: nil
        )
    }
}

struct ContentMentionChipPage: View {
    @Environment(GalleryModel.self) private var model
    @State private var rawMode = false

    var body: some View {
        let profile = model.profile(forPubkey: SHOWCASE_PUBKEY_HEX)
        VStack(spacing: 16) {
            // A hidden avatar owns the claim for the showcase pubkey while this
            // page is visible — same mechanism as `ProfileEmbedPage`, so the
            // mention chip resolves the display name instead of the hex fallback.
            NostrAvatar(pubkey: SHOWCASE_PUBKEY_HEX, size: 0)
                .frame(width: 0, height: 0)
                .clipped()
            RawToggle(rawMode: $rawMode)
            ContentPageFrame(caption: "NostrContentView — live mention resolution") {
                NostrContentView(
                    tree: MentionShowcase.note,
                    mentionLabel: { uri in
                        rawMode
                            ? uri.uri
                            : model.profile(forPubkey: uri.primaryId)?.displayName
                                ?? NostrContentView.defaultMentionLabel(uri)
                    }
                )
                Text("The kernel fetches kind:0 automatically; the app just reads the snapshot.")
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .padding(.top, 4)
            }
            ContentPageFrame(caption: "NostrMentionChip — kernel-backed profile") {
                NostrMentionChip(
                    pubkey: SHOWCASE_PUBKEY_HEX,
                    displayName: profile?.displayName,
                    avatarUrl: profile?.avatarURL
                )
            }
            ContentPageFrame(caption: "NostrMentionChip — reference fallback") {
                NostrMentionChip(
                    pubkey: SHOWCASE_PUBKEY_HEX,
                    displayName: nil
                )
            }
            ContentPageFrame(caption: "NostrMentionChip — no avatar") {
                NostrMentionChip(
                    pubkey: SHOWCASE_PUBKEY_HEX,
                    displayName: profile?.displayName,
                    showsAvatar: false
                )
            }
        }
    }
}

// MARK: - content-minimal

struct ContentMinimalPage: View {
    @Environment(GalleryModel.self) private var model
    @State private var rawMode = false

    var body: some View {
        VStack(spacing: 16) {
            // A hidden avatar owns the claim for the showcase pubkey while this
            // page is visible — same mechanism as `ProfileEmbedPage`, so the
            // minimal-run mention resolves the display name, not the hex fallback.
            NostrAvatar(pubkey: SHOWCASE_PUBKEY_HEX, size: 0)
                .frame(width: 0, height: 0)
                .clipped()
            RawToggle(rawMode: $rawMode)
            ContentPageFrame(caption: "NostrMinimalContentView(runs:)") {
                NostrMinimalContentView(runs: runs)
            }
        }
    }

    private var runs: [NostrContentRun] {
        ShowcaseContent.richTree.nostrMinimalRuns(
            mentionLabel: rawMode
                ? { uri in uri.uri }
                : { uri in
                    model.profile(forPubkey: uri.primaryId)?.displayName
                        ?? NostrContentView.defaultMentionLabel(uri)
                }
        )
    }
}

// MARK: - content-media-grid

struct ContentMediaGridPage: View {
    @Environment(GalleryModel.self) private var model

    var body: some View {
        let imageUrls = relayMediaUrls(from: model.embedHost)
        VStack(spacing: 16) {
            ContentPageFrame(caption: "NostrMediaGrid — relay-backed article media") {
                if imageUrls.isEmpty {
                    Text("Waiting for relay-backed media from the claimed article.")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                } else {
                    NostrMediaGrid(imageUrls: imageUrls)
                }
            }
        }
        .task(id: SHOWCASE_ARTICLE_NADDR) {
            model.embedClaimSink.claim(
                uri: SHOWCASE_ARTICLE_NADDR,
                consumerId: "nmp-gallery-ios.content-media-grid"
            )
        }
        .onDisappear {
            model.embedClaimSink.release(
                uri: SHOWCASE_ARTICLE_NADDR,
                consumerId: "nmp-gallery-ios.content-media-grid"
            )
        }
    }
}

// MARK: - content-quote-card

struct ContentQuoteCardPage: View {
    @Environment(GalleryModel.self) private var model

    var body: some View {
        let quoteModel = relayQuoteModel(from: model)
        VStack(spacing: 16) {
            // The `.task` below claims the quote-card *event* (kind:1 content).
            // The author display name is a separate profile-host claim — owned
            // here by a hidden avatar, mirroring `ProfileEmbedPage`. The author
            // pubkey is statically `SHOWCASE_PUBKEY_HEX`; we claim it directly
            // rather than waiting on the resolved `note.authorPubkey`.
            NostrAvatar(pubkey: SHOWCASE_PUBKEY_HEX, size: 0)
                .frame(width: 0, height: 0)
                .clipped()
            ContentPageFrame(caption: "NostrQuoteCard — rich") {
                NostrQuoteCard(
                    model: quoteModel,
                    variant: .rich
                )
            }
            ContentPageFrame(caption: "NostrQuoteCard — compact") {
                NostrQuoteCard(
                    model: quoteModel,
                    variant: .compact
                )
            }
            ContentPageFrame(caption: "NostrQuoteCard — collapsed") {
                NostrQuoteCard(
                    model: quoteModel,
                    variant: .collapsed
                )
            }
            ContentPageFrame(caption: "NostrQuoteCard — missing") {
                NostrQuoteCard(
                    model: NostrQuoteCardModel(
                        id: SHOWCASE_NOTE_EVENT_ID,
                        unresolvedUri: SHOWCASE_NOTE_NEVENT
                    ),
                    variant: .missing
                )
            }
        }
        .task(id: SHOWCASE_NOTE_NEVENT) {
            model.embedClaimSink.claim(
                uri: SHOWCASE_NOTE_NEVENT,
                consumerId: "nmp-gallery-ios.content-quote-card"
            )
        }
        .onDisappear {
            model.embedClaimSink.release(
                uri: SHOWCASE_NOTE_NEVENT,
                consumerId: "nmp-gallery-ios.content-quote-card"
            )
        }
    }
}

private let showcaseNoteURI = NostrWireUri(
    uri: SHOWCASE_NOTE_NEVENT,
    kind: .event,
    primaryId: SHOWCASE_NOTE_EVENT_ID
)

@MainActor
private func relayMediaUrls(from host: EmbedHost) -> [URL] {
    var values: [String] = []
    if let article = host.envelopeForPrimaryID(SHOWCASE_ARTICLE_PRIMARY_ID),
       case .article(let projection) = article.projection,
       let hero = projection.heroImageUrl {
        values.append(hero)
    }
    if let note = host.envelopeForPrimaryID(SHOWCASE_NOTE_EVENT_ID),
       case .shortNote(let projection) = note.projection {
        values.append(contentsOf: projection.mediaUrls)
    }
    return values.compactMap(URL.init(string:))
}

@MainActor
private func relayQuoteModel(from model: GalleryModel) -> NostrQuoteCardModel {
    guard let envelope = model.embedHost.envelopeForPrimaryID(SHOWCASE_NOTE_EVENT_ID),
          case .shortNote(let note) = envelope.projection
    else {
        return NostrQuoteCardModel(
            id: SHOWCASE_NOTE_EVENT_ID,
            unresolvedUri: SHOWCASE_NOTE_NEVENT
        )
    }
    let profile = model.profile(forPubkey: note.authorPubkey)
    return NostrQuoteCardModel(
        id: note.id,
        unresolvedUri: SHOWCASE_NOTE_NEVENT,
        authorPubkey: note.authorPubkey,
        authorDisplayName: note.authorDisplayName ?? profile?.displayName,
        authorAvatarUrl: (note.authorPictureUrl ?? profile?.pictureUrl).flatMap(URL.init(string:)),
        content: note.content,
        mediaThumbnailUrl: note.mediaUrls.first.flatMap(URL.init(string:)),
        createdAtDisplay: note.createdAt == 0 ? nil : NostrRelativeTime.ago(note.createdAt)
    )
}

File: apps/nmp-gallery/android/app/src/main/kotlin/org/nmp/gallery/gallery/ContentComponentPages.kt
package org.nmp.gallery.gallery

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import org.nmp.gallery.bridge.GalleryModel
import org.nmp.gallery.bridge.GalleryShowcaseReferences
import org.nmp.gallery.bridge.ResolvedEventEnvelopeWire
import org.nmp.gallery.bridge.projectionContentText
import org.nmp.gallery.bridge.projectionLong
import org.nmp.gallery.bridge.projectionString
import org.nmp.gallery.bridge.projectionStrings
import org.nmp.gallery.registry.ContentTreeWire
import org.nmp.gallery.registry.LocalNostrContentRenderer
import org.nmp.gallery.registry.MediaKind
import org.nmp.gallery.registry.NostrContentRenderer
import org.nmp.gallery.registry.NostrContentView
import org.nmp.gallery.registry.NostrMediaGrid
import org.nmp.gallery.registry.NostrMentionChip
import org.nmp.gallery.registry.NostrQuoteCard
import org.nmp.gallery.registry.NostrQuoteCardModel
import org.nmp.gallery.registry.NostrQuoteCardVariant
import org.nmp.gallery.registry.NostrRelativeTime
import org.nmp.gallery.registry.ProfileWire
import org.nmp.gallery.registry.WireNode
import org.nmp.gallery.registry.WireNostrUri
import org.nmp.gallery.registry.WireNostrUriKind
import org.nmp.gallery.registry.defaultMentionLabel

@Composable
fun ContentComponentPage(model: GalleryModel, componentId: String) {
    val profileMap by model.profileMap.collectAsState()
    val resolvedEventEmbeds by model.resolvedEventEmbeds.collectAsState()
    val showcase = model.showcase
    var rawMode by remember { mutableStateOf(false) }

    DisposableEffect(showcase) {
        model.resolveProfileRef(showcase.profile.pubkeyHex, GalleryModel.CONSUMER_ID)
        model.resolveEventRef(showcase.note.uri, GalleryModel.CONSUMER_ID)
        model.resolveEventRef(showcase.article.uri, GalleryModel.CONSUMER_ID)
        onDispose {
            model.releaseProfileRef(showcase.profile.pubkeyHex, GalleryModel.CONSUMER_ID)
            model.releaseEventRef(showcase.note.uri, GalleryModel.CONSUMER_ID)
            model.releaseEventRef(showcase.article.uri, GalleryModel.CONSUMER_ID)
        }
    }

    val renderer = NostrContentRenderer(
        textColor = MaterialTheme.colorScheme.onSurface,
        secondaryTextColor = MaterialTheme.colorScheme.onSurfaceVariant,
        mentionColor = MaterialTheme.colorScheme.primary,
        hashtagColor = MaterialTheme.colorScheme.tertiary,
        linkColor = MaterialTheme.colorScheme.primary,
    )

    val mentionLabel: (WireNostrUri) -> String = { uri ->
        if (rawMode) uri.uri
        else profileMap[uri.primaryId]?.displayName ?: defaultMentionLabel(uri)
    }

    val showsRawToggle = componentId in setOf(
        "content-view", "content-mention-chip", "content-minimal",
    )

    CompositionLocalProvider(LocalNostrContentRenderer provides renderer) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = labelFor(componentId),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            if (showsRawToggle) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        text = if (rawMode) "Raw wire" else "Rendered",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Switch(checked = rawMode, onCheckedChange = { rawMode = it })
                }
            }
            ContentComponentBody(
                componentId = componentId,
                showcase = showcase,
                profileMap = profileMap,
                resolvedEventEmbeds = resolvedEventEmbeds,
                mentionLabel = mentionLabel,
            )
        }
    }
}

@Composable
private fun ContentComponentBody(
    componentId: String,
    showcase: GalleryShowcaseReferences,
    profileMap: Map<String, ProfileWire>,
    resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>,
    mentionLabel: (WireNostrUri) -> String,
) {
    when (componentId) {
        "content-core" -> ContentCoreShowcase(showcase, mentionLabel)
        "content-view" -> ContentViewShowcase(showcase, mentionLabel, resolvedEventEmbeds)
        "content-mention-chip" -> MentionChipShowcase(showcase, profileMap, mentionLabel)
        "content-minimal" -> MinimalContentShowcase(showcase, mentionLabel)
        "content-media-grid" -> MediaGridShowcase(resolvedEventEmbeds)
        "content-quote-card" -> QuoteCardShowcase(showcase, profileMap, resolvedEventEmbeds)
        else -> Text("Unknown content component: $componentId")
    }
}

@Composable
private fun ContentCoreShowcase(
    showcase: GalleryShowcaseReferences,
    mentionLabel: (WireNostrUri) -> String,
) {
    NostrContentView(tree = showcaseTextTree(showcase), mentionLabel = mentionLabel)
}

@Composable
private fun ContentViewShowcase(
    showcase: GalleryShowcaseReferences,
    mentionLabel: (WireNostrUri) -> String,
    resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>,
) {
    NostrContentView(
        tree = showcaseRichTree(showcase),
        mentionLabel = mentionLabel,
        quoteCardProvider = { uri -> quoteCardFor(uri, resolvedEventEmbeds) },
    )
}

@Composable
private fun MentionChipShowcase(
    showcase: GalleryShowcaseReferences,
    profileMap: Map<String, ProfileWire>,
    mentionLabel: (WireNostrUri) -> String,
) {
    val primary = profileMap[showcase.profile.pubkeyHex]
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text("NostrContentView — inline mention", style = MaterialTheme.typography.bodySmall)
        NostrContentView(tree = showcaseMentionTree(showcase), mentionLabel = mentionLabel)
        Text("NostrMentionChip — live kernel-backed", style = MaterialTheme.typography.bodySmall)
        NostrMentionChip(
            pubkey = showcase.profile.pubkeyHex,
            displayName = primary?.displayName,
            avatarUrl = primary?.pictureUrl,
        )
        Text("No avatar variant", style = MaterialTheme.typography.bodySmall)
        NostrMentionChip(
            pubkey = showcase.profile.pubkeyHex,
            displayName = primary?.displayName,
            showsAvatar = false,
        )
    }
}

@Composable
private fun MinimalContentShowcase(
    showcase: GalleryShowcaseReferences,
    mentionLabel: (WireNostrUri) -> String,
) {
    NostrContentView(tree = showcaseShortTree(showcase), mentionLabel = mentionLabel)
}

@Composable
private fun MediaGridShowcase(resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>) {
    val urls = mediaUrlsFromResolvedEventEmbeds(resolvedEventEmbeds)
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("Relay-backed media", style = MaterialTheme.typography.bodySmall)
        if (urls.isEmpty()) {
            Text("Waiting for media from the resolved article.", style = MaterialTheme.typography.bodySmall)
        } else {
            NostrMediaGrid(imageUrls = urls)
        }
    }
}

@Composable
private fun QuoteCardShowcase(
    showcase: GalleryShowcaseReferences,
    profileMap: Map<String, ProfileWire>,
    resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>,
) {
    val noteUri = noteUri(showcase)
    val quoteModel = quoteCardFor(noteUri, resolvedEventEmbeds)?.let { model ->
        val profile = model.authorPubkey?.let { profileMap[it] }
        model.copy(
            authorDisplayName = profile?.displayName,
            authorAvatarUrl = profile?.pictureUrl,
        )
    } ?: NostrQuoteCardModel(id = showcase.note.primaryId, unresolvedUri = showcase.note.uri)
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("Rich", style = MaterialTheme.typography.bodySmall)
        NostrQuoteCard(model = quoteModel, variant = NostrQuoteCardVariant.Rich)
        Text("Compact", style = MaterialTheme.typography.bodySmall)
        NostrQuoteCard(model = quoteModel, variant = NostrQuoteCardVariant.Compact)
        Text("Collapsed", style = MaterialTheme.typography.bodySmall)
        NostrQuoteCard(model = quoteModel, variant = NostrQuoteCardVariant.Collapsed)
        Text("Missing", style = MaterialTheme.typography.bodySmall)
        NostrQuoteCard(model = quoteModel.copy(unresolvedUri = showcase.note.uri), variant = NostrQuoteCardVariant.Missing)
    }
}

// ── Relay-reference content trees ────────────────────────────────────────────

private fun showcaseMentionTree(showcase: GalleryShowcaseReferences): ContentTreeWire {
    val nodes = listOf<WireNode>(
        WireNode.Text("Hey "),
        WireNode.Mention(
            uri = profileUri(showcase),
        ),
        WireNode.Paragraph(children = listOf(0u, 1u)),
    )
    return ContentTreeWire(nodes = nodes, roots = listOf(2u))
}

private fun showcaseTextTree(showcase: GalleryShowcaseReferences): ContentTreeWire {
    val nodes = listOf<WireNode>(
        WireNode.EventRef(noteUri(showcase)),
        WireNode.Paragraph(children = listOf(0u)),
    )
    return ContentTreeWire(nodes = nodes, roots = listOf(1u))
}

private fun showcaseShortTree(showcase: GalleryShowcaseReferences): ContentTreeWire {
    val nodes = listOf<WireNode>(
        WireNode.Mention(profileUri(showcase)),
        WireNode.Paragraph(children = listOf(0u)),
    )
    return ContentTreeWire(nodes = nodes, roots = listOf(1u))
}

private fun showcaseRichTree(showcase: GalleryShowcaseReferences): ContentTreeWire {
    val nodes = listOf<WireNode>(
        // 0
        WireNode.Text("Relay note "),
        // 1
        WireNode.Mention(
            uri = profileUri(showcase),
        ),
        // 2
        WireNode.Text(", "),
        // 3
        WireNode.EventRef(noteUri(showcase)),
        // 4
        WireNode.Text(" "),
        // 5
        WireNode.Paragraph(children = listOf(0u, 1u, 2u, 3u, 4u)),
        // 6
    )
    return ContentTreeWire(
        nodes = nodes,
        roots = listOf(5u),
    )
}

private fun noteUri(showcase: GalleryShowcaseReferences) = WireNostrUri(
    uri = showcase.note.uri,
    kind = WireNostrUriKind.Event,
    primaryId = showcase.note.primaryId,
)

private fun profileUri(showcase: GalleryShowcaseReferences) = WireNostrUri(
    uri = "nostr:${showcase.profile.npub}",
    kind = WireNostrUriKind.Profile,
    primaryId = showcase.profile.pubkeyHex,
)

private fun quoteCardFor(
    uri: WireNostrUri,
    resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>,
): NostrQuoteCardModel? {
    val event = resolvedEventEmbeds[uri.primaryId] ?: return null
    return NostrQuoteCardModel(
        id = event.projectionString("id") ?: event.primaryId,
        unresolvedUri = uri.uri,
        authorPubkey = event.projectionString("authorPubkey"),
        content = event.projectionString("content") ?: event.projectionContentText(),
        mediaThumbnailUrl = mediaUrls(event).firstOrNull(),
        createdAtDisplay = event.projectionLong("createdAt")
            ?.takeIf { it > 0L }
            ?.let { NostrRelativeTime.ago(it) },
    )
}

private fun mediaUrlsFromResolvedEventEmbeds(
    resolvedEventEmbeds: Map<String, ResolvedEventEnvelopeWire>,
): List<String> = resolvedEventEmbeds.values.flatMap(::mediaUrls).distinct()

private fun mediaUrls(event: ResolvedEventEnvelopeWire): List<String> {
    val projected = event.projectionStrings("mediaUrls") +
        listOfNotNull(event.projectionString("heroImageUrl"))
    val inline = event.projectionString("content").orEmpty()
        .split(Regex("\\s+"))
        .filter(::looksLikeMedia)
    return (projected + inline).distinct()
}

private fun looksLikeMedia(value: String): Boolean {
    val lower = value.lowercase()
    return (lower.startsWith("http://") || lower.startsWith("https://")) &&
        listOf(".jpg", ".jpeg", ".png", ".gif", ".webp").any { lower.contains(it) }
}

private fun labelFor(componentId: String): String = when (componentId) {
    "content-core" -> "ContentTreeWire (flat arena)"
    "content-view" -> "NostrContentView — live mention resolution"
    "content-mention-chip" -> "NostrMentionChip — kernel-backed profiles"
    "content-minimal" -> "NostrMinimalContentView"
    "content-media-grid" -> "NostrMediaGrid"
    "content-quote-card" -> "NostrQuoteCard — kernel-backed author"
    else -> componentId
}
